File: solutions/_15_3sum/solution.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        n = len(nums)
        for i in range(n - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            L = i + 1
            R = n - 1
            while L < R:
                current_sum = nums[i] + nums[L] + nums[R]
                if current_sum < 0:
                    L += 1
                elif current_sum > 0:
                    R -= 1
                else:
                    res.append([nums[i], nums[L], nums[R]])
                    while L < R and nums[L] == nums[L + 1]:
                        L += 1
                    while L < R and nums[R] == nums[R - 1]:
                        R -= 1
                    L += 1
                    R -= 1
        return res


# https://leetcode.com/problems/3sum/

if __name__ == "__main__":
    s = Solution()
    print(s.threeSum([-1, 0, 1, 2, -1, -4]))  # [[-1, -1, 2],

File: solutions/_15_3sum/solution.cpp
#include <vector>
using namespace std;

class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        vector<vector<int>> result;
        sort(nums.begin(), nums.end());
        int n = nums.size();

        for (int i = 0; i < n - 2; i++)
        {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            int L = i + 1;
            int R = n - 1;

            while (L < R)
            {
                int sum = nums[i] + nums[L] + nums[R];
                if (sum < 0)
                    L++;
                else if (sum > 0)
                    R--;
                else
                {
                    result.push_back({nums[i], nums[L], nums[R]});
                    while (L < R && nums[L] == nums[L + 1])
                        L++;
                    while (L < R && nums[R] == nums[R - 1])
                        R--;
                    L++;
                    R--;
                }
            }
        }
        return result;
    }
};

/*
https://leetcode.com/problems/3sum/
*/